ark/type/parser/objectLiteral.ts

```diff
@@ -55,23 +110,31 @@ export const parseObjectLiteral = (def: Dict, ctx: ParseContext) => {
 }
 
 export type inferObjectLiteral<def extends object, $, args> = evaluate<
-	{
-		// since def is a const parameter, we remove the readonly modifier here
-		// support for builtin readonly tracked here:
-		// https://github.com/arktypeio/arktype/issues/808
-		-readonly [k in keyof def as nonOptionalKeyFrom<
-			k,
-			def[k],
-			$,
-			args
-		>]: inferDefinition<def[k], $, args>
-	} & {
-		-readonly [k in keyof def as optionalKeyFrom<k, def[k]>]?: inferDefinition<
-			def[k] extends OptionalValue<infer inner> ? inner : def[k],
-			$,
-			args
-		>
-	}
+	merge<
+		def extends { readonly "...": infer spreadDef }
```

Comment:
So we don't want to run this merge utility every time we have an object definition as it will be very expensive.

There's a few options which we'd want to test the performance/accuracy of a bit.

1) Split this into multiple generics that would compute the inferred types for optional and required keys *including* those that came from a spread, if there is one.
2) Check if  `...` is a keyof def before merging anything.

Reply:
Does 18c8bdf do what you'd like?